### ISSUE TITLE
Nuxt does not clean up CSS after navigation (and injects CSS pre-navigation)

### ISSUE THREAD

**Issue by matthew-dean** [nuxt/nuxt#22817]
### Environment

- Operating System: `Darwin`
- Node Version:     `v16.19.0`
- Nuxt Version:     `3.6.5`
- Nitro Version:    `2.5.2`
- Package Manager:  `pnpm@8.3.1`
- Builder:          `vite`
- User Config:      `extends`, `srcDir`, `build`, `hooks`, `runtimeConfig`
- Runtime Modules:  `-`
- Build Modules:    `-`

### Reproduction

https://stackblitz.com/edit/nuxt-starter-l1dvnf?file=app.vue

### Describe the bug

There's actually two manifestations of this bug.

First bug:
- Given two pages, A & B, and two layouts, X and Y
- Page A uses Template X
- Page B uses Template Y
- When navigating to page A, it will mount the CSS for Template X
- When navigating to page B, it will mount the CSS for Template Y
- However, when navigating **_back_** to page A, it will not _remove_ the CSS for template Y.
- **Bug:** This means, effectively, that page A's rendering is unstable, and will render differently depending on preceding navigation.

Second bug:
- Given the above, if you use pre-fetching with `NuxtLink` (`<NuxtLink>` without `noPrefetch`), then the CSS for template Y will be injected into the page when navigating to page A, even though it has no association with Template Y.
- In other words, Nuxt eagerly injects CSS un-associated with the current page

Note that the first bug occurs whether or not you use `noPrefetch`.

### Additional context

I originally thought this bug was associated with / the same bug as https://github.com/nuxt/nuxt/issues/3877, but was told that was not the case, so filing this new issue. I also thought it might be the exact same as [this issue](https://github.com/nuxt/nuxt/issues/13778), but that issue is closed?

**Comment by danielroe** [comment]
apologies for the delay in looking at this.

**Comment by galaxyblur** [comment]
I'm seeing this issue when not using layouts -- simply two different pages. When I `@import` within the `<style></style>` block of a `page`, Page A will keep the css files loaded from Page B after navigating B->A.

**Comment by manniL** [comment]
@matthew-dean No, absolutely not! 😋

> Page A will keep the css files loaded from Page B after navigating B->A.

This is expected and you can use `<style scoped>` (Scoped Styles) to avoid this. No styles will be "unloaded" when navigating. If your page has global styles that alter the whole application (non-scoped ones), they will stay throughout the user journey after being loaded unless somehow handled in userland ☺️

**Comment by danielroe** [comment]
I agree with @manniL's take. You should not rely on whether a stylesheet is or isn't loaded. Nuxt may preload a stylesheet when the browser is idle, for example, to ensure that a page loads quickly and doesn't rely on additional network requests.

Vite may also bundle multiple pages' or layouts' CSS into a single file if they are very small.

These are implementation details.

Instead you should use scoped CSS, CSS modules, or unique selectors to ensure there are no conflicts. This is a matter of (IMO) best practice anyway.

**Comment by danielroe** [comment]
The behaviour here is not Nuxt's but Vite's; I would create a proposal or feature request there, that we can then benefit from in Nuxt.

**Comment by wavedeck** [comment]
> > Page A will keep the css files loaded from Page B after navigating B->A.
>
> IMO This is expected and you can use `<style scoped>`

While this is true for styles applied to individual pages, you cannot scope styles inside a layout, as this will not apply them to the pages using the layout.. so the only real workaround is to encapsulate the layout styles in another, parent selector
(e.g. ".layout-a body" and ".layout-b body") or using completely unique selectors.

this issue drove me insane when i was trying to reduce css bundle size by splitting them into different layouts. The second you use NuxtLink without disabling prefetching, the styles will get combined again and can cause unexpected behaviour.

It also didn't help that this issue was closed, since it makes it more effort to find.

In general, I'm distressed that almost all of the major JavaScript SSR frameworks out there happen to have so many gotcha's when handling css. It's very hard to keep control and regularly lures me back to traditional SSR, where you have full control of the HTML for each page.

**Comment by manniL** [comment]
> While this is true for styles applied to individual pages, you cannot scope styles inside a layout, as this will not apply them to the pages using the layout.. 

You can use the `deep` pseudoclass for that :)

**Comment by slps970093** [comment]
> The behaviour here is not Nuxt's but Vite's; I would create a [proposal or feature request there](https://github.com/vitejs/vite), that we can then benefit from in Nuxt.

Is there a link to this question? I wonder if it's already here

**Comment by danielroe** [comment]
Reopened to investigate.

**Comment by metkm** [comment]
> Reopened to investigate.

I think something like this would be useful when using view-transitions. When you need to animate something with different aspect ratios you should apply different `object-fit` property depending on the direction.

For example, when you go from wider image to smaller image
```
::view-transition-old(cover) {
  object-fit: cover;
}

::view-transition-new(cover) {
  object-fit: contain;
}
```
vice versa
```
::view-transition-old(cover) {
  object-fit: contain;
}

::view-transition-new(cover) {
  object-fit: cover;
}
```

**Comment by GrantGryczan** [comment]
> IMO This is expected and you can use <style scoped> (Scoped Styles) to avoid this.

@manniL You cannot use scoped styles to avoid this when the selector you're trying to set styles for cannot be scoped. For example, if I want one page to have a different background color than another page, I'd have to jump through hoops and use a script to do that, even though intuitively I should just be able to use a `body` selector (or `:global(body)` in scoped CSS). This behavior not only leads to memory leaks but is also not expected or beneficial as far as I can tell.

**Comment by Kreeg** [comment]
This bug still exist in Nuxt 4. Any progress on that?

**Comment by Mateleo** [comment]
@Kreeg Yes, nothing changed because it's a Vite related issue, not directly nuxt.
All three of these have to be true at once to trigger this bug:

1. You write global (non-scoped) styles inside a page or layout, a plain `<style>` block, not `<style scoped>`, targeting a shared selector like `body`, `html`, `:root`, or a generic `class`.
2. Two different routes set the same global selector to different values (e.g. layout X does` body { background: red }`, layout Y does `body { background: blue }`).
3. You navigate between them.

I have no idea how to fix this for Nuxt. Maybe an `experimental.cleanupStylesOnNavigation`
An ideas ? @OrbisK @Flo0806

**Comment by Flo0806** [comment]
This issue is on my list - but it is tricky. I'll have a look tomorrow (or today in Germany :-) )

**Comment by Aryakoste** [comment]
@Flo0806 @danielroe  I have submitted a PR for this.
